Panic log

mobius_client_linux_amd64$ ./mobius-hotline-client -config mobius-client-config.yaml 
panic: runtime error: index out of range [22] with length 18 [recovered]
        panic: runtime error: index out of range [22] with length 18

goroutine 1 [running]:
github.com/rivo/tview.(*Application).Run.func1()
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/application.go:243 +0x4d
panic({0x718ec0, 0xc00001e480})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/jhalter/mobius/hotline.serverScanner({0xc0006a2588?, 0x480?, 0xc00063df38?}, 0x0?)
        /home/circleci/project/hotline/tracker.go:158 +0x86
bufio.(*Scanner).Scan(0xc00074d7a0)
        /usr/local/go/src/bufio/scan.go:146 +0xbf
github.com/jhalter/mobius/hotline.GetListing({0xc000536060?, 0x0?})
        /home/circleci/project/hotline/tracker.go:126 +0x45e
github.com/jhalter/mobius/hotline.(*UI).getTrackerList(0xc0005b6240)
        /home/circleci/project/hotline/ui.go:107 +0x49
github.com/jhalter/mobius/hotline.(*UI).Start.func3()
        /home/circleci/project/hotline/ui.go:474 +0x25
github.com/rivo/tview.(*List).InputHandler.func1(0xc00074d9d0?, 0x56455f?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/list.go:619 +0xcf
github.com/rivo/tview.(*Box).WrapInputHandler.func1(0xc00011ca90?, 0xc00074da10?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/box.go:167 +0x53
github.com/rivo/tview.(*Flex).InputHandler.func1(0xc00074da50?, 0x55e8ff?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/flex.go:251 +0xd9
github.com/rivo/tview.(*Box).WrapInputHandler.func1(0xc0005185d0?, 0xc00074da90?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/box.go:167 +0x53
github.com/rivo/tview.(*Flex).InputHandler.func1(0xc00074dad0?, 0x55e8ff?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/flex.go:251 +0xd9
github.com/rivo/tview.(*Box).WrapInputHandler.func1(0xc0005184b0?, 0x7f4c610af7d0?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/box.go:167 +0x53
github.com/rivo/tview.(*Pages).InputHandler.func1(0x40ea67?, 0x10?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/pages.go:311 +0xa7
github.com/rivo/tview.(*Box).WrapInputHandler.func1(0xc0005182d0?, 0xc00074dbd8?)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/box.go:167 +0x53
github.com/rivo/tview.(*Application).Run(0xc0005261c0)
        /home/circleci/go/pkg/mod/github.com/rivo/tview@v0.0.0-20220307222120-9994674d60a8/application.go:335 +0x77d
github.com/jhalter/mobius/hotline.(*UI).Start(0xc0005b6240)
        /home/circleci/project/hotline/ui.go:508 +0xd45
main.main()
        /home/circleci/project/cmd/mobius-hotline-client/main.go:76 +0x8e5
        
Was using Kubuntu 22.

--- a/hotline/tracker.go
+++ b/hotline/tracker.go
@@ -125,7 +125,10 @@ func GetListing(addr string) ([]ServerRecord, error) {
 	for {
 		scanner.Scan()
 		var srv ServerRecord
-		_, _ = srv.Read(scanner.Bytes())
+		_, err = srv.Read(scanner.Bytes())
+		if err != nil {
+			return nil, err
+		}
 
 		servers = append(servers, srv)
 		if len(servers) == totalSrv {
@@ -146,16 +149,28 @@ func GetListing(addr string) ([]ServerRecord, error) {
 // 00000050  4f 54 4c 2e 63 6f 6d 3a  35 35 30 30 2d 4f 3a b2  |OTL.com:5500-O:.|
 // 00000060  15 7c 00 00 00 00 08 53  65 6e 65 63 74 75 73 20  |.|.....Senectus |
 func serverScanner(data []byte, _ bool) (advance int, token []byte, err error) {
+	// The name length field is the 11th byte of the server record.  If we don't have that many bytes,
+	// return nil token so the Scanner reads more data and continues scanning.
 	if len(data) < 10 {
 		return 0, nil, nil
 	}
 
 	// A server entry has two variable length fields: the name and description.
-	// To get the token length, we first need the name length from the 10th byte:
+	// To get the token length, we first need the name length from the 10th byte
 	nameLen := int(data[10])
+
+	// The description length field is at the 12th + nameLen byte of the server record.
+	// If we don't have that many bytes, return nil token so the Scanner reads more data and continues scanning.
+	if len(data) < 11+nameLen {
+		return 0, nil, nil
+	}
 
 	// Next we need the description length from the 11+nameLen byte:
 	descLen := int(data[11+nameLen])
+
+	if len(data) < 12+nameLen+descLen {
+		return 0, nil, nil
+	}
 
 	return 12 + nameLen + descLen, data[0 : 12+nameLen+descLen], nil
 }

--- a/hotline/tracker_test.go
+++ b/hotline/tracker_test.go
@@ -139,6 +139,44 @@ func Test_serverScanner(t *testing.T) {
 			wantToken:   []byte(nil),
 			wantErr:     assert.NoError,
 		},
+		{
+			name: "when nameLen exceeds provided data",
+			args: args{
+				data: []byte{
+					0x18, 0x05, 0x30, 0x63, // IP Addr
+					0x15, 0x7c, // Port
+					0x00, 0x02, // UserCount
+					0x00, 0x00, // ??
+					0xff,             // Name Len
+					0x54, 0x68, 0x65, // Name
+					0x03,             // Desc Len
+					0x54, 0x54, 0x54, // Description
+				},
+				atEOF: false,
+			},
+			wantAdvance: 0,
+			wantToken:   []byte(nil),
+			wantErr:     assert.NoError,
+		},
+		{
+			name: "when description len exceeds provided data",
+			args: args{
+				data: []byte{
+					0x18, 0x05, 0x30, 0x63, // IP Addr
+					0x15, 0x7c, // Port
+					0x00, 0x02, // UserCount
+					0x00, 0x00, // ??
+					0x03,             // Name Len
+					0x54, 0x68, 0x65, // Name
+					0xff,             // Desc Len
+					0x54, 0x54, 0x54, // Description
+				},
+				atEOF: false,
+			},
+			wantAdvance: 0,
+			wantToken:   []byte(nil),
+			wantErr:     assert.NoError,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/hotline/ui.go
+++ b/hotline/ui.go
@@ -102,12 +102,7 @@ func (ui *UI) showBookmarks() *tview.List {
 	return list
 }
 
-func (ui *UI) getTrackerList() *tview.List {
-	listing, err := GetListing(ui.HLClient.Pref.Tracker)
-	if err != nil {
-		// TODO
-	}
-
+func (ui *UI) getTrackerList(servers []ServerRecord) *tview.List {
 	list := tview.NewList()
 	list.SetInputCapture(func(event *tcell.EventKey) *tcell.EventKey {
 		if event.Key() == tcell.KeyEsc {
@@ -117,14 +112,12 @@ func (ui *UI) getTrackerList() *tview.List {
 	})
 	list.Box.SetBorder(true).SetTitle("| Servers |")
 
-	shortcut := 97 // rune for "a"
-	for i, srv := range listing {
-		addr := srv.Addr()
-		srvName := srv.Name
+	const shortcut = 97 // rune for "a"
+	for i, srv := range servers {
 		list.AddItem(string(srv.Name), string(srv.Description), rune(shortcut+i), func() {
 			ui.Pages.RemovePage("joinServer")
 
-			newJS := ui.renderJoinServerForm(string(srvName), addr, GuestAccount, "", trackerListPage, false, true)
+			newJS := ui.renderJoinServerForm(string(srv.Name), srv.Addr(), GuestAccount, "", trackerListPage, false, true)
 
 			ui.Pages.AddPage("joinServer", newJS, true, true)
 			ui.Pages.ShowPage("joinServer")
@@ -411,7 +404,7 @@ func (ui *UI) renderServerUI() *tview.Flex {
 							newsPostTextArea.SetText(curTxt)
 						}
 					default:
-						_, _ = fmt.Fprintf(newsPostTextArea, string(event.Rune()))
+						_, _ = fmt.Fprint(newsPostTextArea, string(event.Rune()))
 					}
 				}
 
@@ -468,7 +461,20 @@ func (ui *UI) Start() {
 			ui.Pages.AddAndSwitchToPage("bookmarks", ui.showBookmarks(), true)
 		}).
 		AddItem("Browse Tracker", "", 't', func() {
-			ui.trackerList = ui.getTrackerList()
+			listing, err := GetListing(ui.HLClient.Pref.Tracker)
+			if err != nil {
+				errMsg := fmt.Sprintf("Error fetching tracker results:\n%v", err)
+				errModal := tview.NewModal()
+				errModal.SetText(errMsg)
+				errModal.AddButtons([]string{"Cancel"})
+				errModal.SetDoneFunc(func(buttonIndex int, buttonLabel string) {
+					ui.Pages.RemovePage("errModal")
+				})
+				ui.Pages.RemovePage("joinServer")
+				ui.Pages.AddPage("errModal", errModal, false, true)
+				return
+			}
+			ui.trackerList = ui.getTrackerList(listing)
 			ui.Pages.AddAndSwitchToPage("trackerList", ui.trackerList, true)
 		}).
 		AddItem("Settings", "", 's', func() {

